pkg/deployment/cluster_scaling_integration.go

```diff
@@ -253,3 +252,70 @@ func (ci *clusterScalingIntegration) updateClusterServerCount(ctx context.Contex
 	ci.lastNumberOfServers.DBServers = &dbserverCount
 	return safeToAskCluster, nil
 }
+
+// GetLastNumberOfServers returns the last number of servers
+func (ci *clusterScalingIntegration) GetLastNumberOfServers() arangod.NumberOfServers {
+	ci.lastNumberOfServers.mutex.Lock()
+	defer ci.lastNumberOfServers.mutex.Unlock()
+
+	return ci.lastNumberOfServers.NumberOfServers
+}
+
+// DisableScalingCluster disables scaling DBservers and coordinators
+func (ci *clusterScalingIntegration) DisableScalingCluster() error {
+	ci.scaleEnabled.mutex.Lock()
+	defer ci.scaleEnabled.mutex.Unlock()
+
+	// Turn off scaling DBservers and coordinators in arangoDB for the UI
+	ctx := context.Background()
+	if err := ci.depl.SetNumberOfServers(ctx, nil, nil); err != nil {
+		return maskAny(err)
+	}
+
+	ci.scaleEnabled.enabled = false
+	return nil
+}
+
+// EnableScalingCluster enables scaling DBservers and coordinators
+func (ci *clusterScalingIntegration) EnableScalingCluster() error {
+	ci.scaleEnabled.mutex.Lock()
+	defer ci.scaleEnabled.mutex.Unlock()
+
+	if ci.scaleEnabled.enabled {
+		return nil
+	}
+
+	ctx := context.Background()
+	lastNumberOfServers := ci.GetLastNumberOfServers()
+	err := ci.depl.SetNumberOfServers(ctx, lastNumberOfServers.Coordinators, lastNumberOfServers.DBServers)
+	if err != nil {
+		return maskAny(err)
+	}
+	ci.scaleEnabled.enabled = true
+	return nil
+}
+
+// IsScalingEnabled checks if scaling is allowed.
+func (ci *clusterScalingIntegration) IsScalingEnabled() bool {
+	ci.scaleEnabled.mutex.Lock()
+
+	if ci.scaleEnabled.enabled {
+		ci.scaleEnabled.mutex.Unlock()
+		return true
+	}
+
+	// Scaling is disabled
+	status, _ := ci.depl.GetStatus()
+	if !status.Plan.Empty() {
+		// Some Plan exists and scaling can not be enabled
+		ci.scaleEnabled.mutex.Unlock()
+		return false
+	}
+
+	// Scaling should be enabled because there is no Plan. It can happen when the action with enabling
+	ci.scaleEnabled.mutex.Unlock()
+	if ci.EnableScalingCluster() != nil {
```

Comment:
Here is a race! After line 316 in theory a new plan could be build and the `DisableScalingAction` could already been executed. Then this code would undo this change.

Reply:
changed